### ISSUE TITLE
mseed crate doesn't support to create MSRecord by caller

### ISSUE THREAD

**Issue by BIBIN-EUGINE** [damb/mseed#14]
Hi @damb , I have come across with using `mseed::pack_header2(msr: &MSRecord, buf: &mut [u8])` , but I am not sure how to create a `MSRecord` as it's a pointer to MS3Record and doesn't provided any API to initialize it.

It would be nice if there is a example for how to use this API for the caller.

Thanks

**Comment by damb** [comment]
I guess this is somehow related to #13.

There is a reason why there isn't an example, yet. At the time being, this is simply not possible. However, while designing and implementing the API I knew already about these kind of issues.

Currently, `MSRecord` can be instantiated only from a miniSEED record buffer. With this approach a user of `MSRecord` can be sure that it has always a valid and fully initialized instance of `MSRecord`. No invariants are required. On the other hand, we loose some flexibility.

Now to your problem. Basically, I see two approaches to solve this issue.

1. Trade simplicity against flexibility. Allow to partially initialize `MSRecord` which requires several invariants and a complete rework of the `MSRecord` API.
2. Providing overloads for both `pack_header2()` and `pack_header3()` which allow packing headers from `mseed::PackInfo`.

For the sake of simplicity, I'd prefer to implement the latter approach. However, there may be reasons I'm still not aware of and in the long term it might be better to have more flexibility.

What do you think?

**Comment by BIBIN-EUGINE** [comment]
> > I guess this is somehow related to #13.
> > There is a reason why there isn't an example, yet. At the time being, this is simply not possible. However, while designing and implementing the API I knew already about these kind of issues.
> > Currently, `MSRecord` can be instantiated only from a miniSEED record buffer. With this approach a user of `MSRecord` can be sure that it has always a valid and fully initialized instance of `MSRecord`. No invariants are required. On the other hand, we loose some flexibility.
> > Now to your problem. Basically, I see two approaches to solve this issue.
> > 
> > 1. Trade simplicity against flexibility. Allow to partially initialize `MSRecord` which requires several invariants and a complete rework of the `MSRecord` API.
> > 2. Providing overloads for both `pack_header2()` and `pack_header3()` which allow packing headers from `mseed::PackInfo`.
> > 
> > For the sake of simplicity, I'd prefer to implement the latter approach. However, there may be reasons I'm still not aware of and in the long term it might be better to have more flexibility.
> > What do you think?
> > I agree, overloading it with `mseed::PackInfo` make it easy to implement for now. However, it would be nice to refactor the `MSRecord` API to bring more flexbiltiy into it.

**Comment by BIBIN-EUGINE** [comment]
> I guess this is somehow related to #13.
> 
> There is a reason why there isn't an example, yet. At the time being, this is simply not possible. However, while designing and implementing the API I knew already about these kind of issues.
> 
> Currently, `MSRecord` can be instantiated only from a miniSEED record buffer. With this approach a user of `MSRecord` can be sure that it has always a valid and fully initialized instance of `MSRecord`. No invariants are required. On the other hand, we loose some flexibility.
> 
> Now to your problem. Basically, I see two approaches to solve this issue.
> 
> 1. Trade simplicity against flexibility. Allow to partially initialize `MSRecord` which requires several invariants and a complete rework of the `MSRecord` API.
> 2. Providing overloads for both `pack_header2()` and `pack_header3()` which allow packing headers from `mseed::PackInfo`.
> 
> For the sake of simplicity, I'd prefer to implement the latter approach. However, there may be reasons I'm still not aware of and in the long term it might be better to have more flexibility.
> 
> What do you think?

I agree, overloading it with mseed::PackInfo make it easy to implement for now. However, it would be nice to refactor the MSRecord API to bring more flexbiltiy into it
